SalesforceCommerceCloud/pwa-kit
# Fix Mono Repo Package Publishing

## Problem

For eons Mobify has put up with some really crumby import statements with respect to the SDK. The import statements look like this example `import {getAppOrigin} from 'pwa-kit-react-sdk/dist/utils/url'`. The point of contention is the `dist` part of the path. Its just ugly and  something that you wouldn't see in a library worth its salt. This was never fixed in the past as we were too scared to make breaking changes.

## Solution

The reason this happens is that we are deploying from the directory where the `package.json` files lies which includes the `dist` folder. To get around this, you have to run the publish command from the `dist` folder. But before that can work, you have to copy the `package.json` into it. 

This is partially accomplished by using the `--contents <dir>` option provided by lerna. The only downside is that all packages that are "public" will have to have the same folder structure aka have a `dist` folder. To us, this means that the `pwa-kit-create-app` will have to now have a dist folder.

 **GUS**: [W-9655228](https://gus.lightning.force.com/lightning/_classic/%2Fa07AH000000SBGTYA4)
 **Linked PRs**: None

## How to test-drive this PR
- Ensure the PWA still starts are works as expected within the monorepo.
- Ensure a generated project works as intended. Run this command to generate a project. **NOTE: This isn't working until 2 todo items below are done, but I'm looking for input on them.** 
```
GENERATOR_PRESET=test-project node packages/pwa-kit-create-app/scripts/create-mobify-app-dev.js --outputDir /tmp/project
```

## Changes
- Add new script to copy package files into dist and package.json.
- Rewrite the imports for the PWA.
- Add package.json publishConfig entry for SDK.
- Adjust package.json files entries for sdk and pwa.

## Todos
- [x] Fix running `pwa` project locally.

👆🏻👆🏻👆🏻 **NOTE: Looking for suggestions on first 2 items in the above list. Using a file replacement command just seems like bad form.** - Complete
